Rolling update test is failing when partition scaling is enabled

### <!-- Component -->

<!-- Zeebe- -->

### <!-- Affected version -->

<!-- -8.8 -->

### <!-- Severity -->

<!-- High- -->

### <!-- Likelihood -->

<!-- High_ -->

### Description

RollingUpdateTest on stable 8.8 branch is failing when trying to the merge the PR that enables partition scaling by default. 

https://github.com/camunda/camunda/actions/runs/17648076698?pr=37988 

```
[2025-09-15 07:33:21.517] [zb-actors-1] [{actor-name=Startup, actor-scheduler=Broker-1}] WARN 
	io.camunda.zeebe.broker.system - Aborting startup process due to exception during step Cluster Topology Manager
io.camunda.zeebe.dynamic.config.serializer.DecodingFailed: com.google.protobuf.InvalidProtocolBufferException: Protocol message contained an invalid tag (zero).
	at io.camunda.zeebe.dynamic.config.serializer.ProtoBufSerializer.decodeClusterTopology(ProtoBufSerializer.java:133)
	at io.camunda.zeebe.dynamic.config.PersistedClusterConfiguration.readFromFile(PersistedClusterConfiguration.java:107)
	at io.camunda.zeebe.dynamic.config.PersistedClusterConfiguration.ofFile(PersistedClusterConfiguration.java:60)
	at io.camunda.zeebe.dynamic.config.ClusterConfigurationManagerService.<init>(ClusterConfigurationManagerService.java:87)
	at io.camunda.zeebe.broker.partitioning.topology.DynamicClusterConfigurationService.getClusterTopologyManagerService(DynamicClusterConfigurationService.java:169)
	at io.camunda.zeebe.broker.partitioning.topology.DynamicClusterConfigurationService.start(DynamicClusterConfigurationService.java:66)
	at io.camunda.zeebe.broker.bootstrap.ClusterConfigurationManagerStep.startup(ClusterConfigurationManagerStep.java:41)
	at io.camunda.zeebe.broker.bootstrap.ClusterConfigurationManagerStep.startup(ClusterConfigurationManagerStep.java:19)
	at io.camunda.zeebe.broker.bootstrap.BrokerStepMetricDecorator.callDelegateAndUpdateTimer(BrokerStepMetricDecorator.java:62)
	at io.camunda.zeebe.broker.bootstrap.BrokerStepMetricDecorator.startup(BrokerStepMetricDecorator.java:44)
	at io.camunda.zeebe.broker.bootstrap.BrokerStepMetricDecorator.startup(BrokerStepMetricDecorator.java:25)
	at io.camunda.zeebe.scheduler.startup.StartupProcess.proceedWithStartupSynchronized(StartupProcess.java:162)
	at io.camunda.zeebe.scheduler.startup.StartupProcess.lambda$proceedWithStartupSynchronized$3(StartupProcess.java:170)
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:28)
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:86)
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:43)
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:125)
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:127)
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:104)
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:224)
Caused by: com.google.protobuf.InvalidProtocolBufferException: Protocol message contained an invalid tag (zero).
	at com.google.protobuf.InvalidProtocolBufferException.invalidTag(InvalidProtocolBufferException.java:110)
	at com.google.protobuf.CodedInputStream$ArrayDecoder.readTag(CodedInputStream.java:649)
	at io.camunda.zeebe.dynamic.config.protocol.Topology$RequestHandling$Builder.mergeFrom(Topology.java:5007)
	at io.camunda.zeebe.dynamic.config.protocol.Topology$RequestHandling$Builder.mergeFrom(Topology.java:4873)
	at com.google.protobuf.CodedInputStream$ArrayDecoder.readMessage(CodedInputStream.java:868)
	at io.camunda.zeebe.dynamic.config.protocol.Topology$RoutingState$Builder.mergeFrom(Topology.java:2902)
	at io.camunda.zeebe.dynamic.config.protocol.Topology$RoutingState$Builder.mergeFrom(Topology.java:2748)
	at com.google.protobuf.CodedInputStream$ArrayDecoder.readMessage(CodedInputStream.java:868)
	at io.camunda.zeebe.dynamic.config.protocol.Topology$ClusterTopology$Builder.mergeFrom(Topology.java:1790)
	at io.camunda.zeebe.dynamic.config.protocol.Topology$ClusterTopology$1.parsePartialFrom(Topology.java:2385)
	at io.camunda.zeebe.dynamic.config.protocol.Topology$ClusterTopology$1.parsePartialFrom(Topology.java:2377)
	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:111)
	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:126)
	at com.google.protobuf.AbstractParser.parseFrom(AbstractParser.java:25)
	at io.camunda.zeebe.dynamic.config.protocol.Topology$ClusterTopology.parseFrom(Topology.java:1477)
	at io.camunda.zeebe.dynamic.config.serializer.ProtoBufSerializer.decodeClusterTopology(ProtoBufSerializer.java:128)
	... 19 more
```

### Steps to reproduce

Run RollingUpdateTest `shouldPerformRollingUpdate` with partition scaling enabled  by setting the env var `ZEEBE_BROKER_EXPERIMENTAL_FEATURES_ENABLEPARTITIONSCALING`


### Current behavior

.

### Expected behavior

.

### Workaround

Most likely shutting down the whole cluster and restarting it on the new version works. This hypothesis is based on the fact that other update tests which are not rolling passes.

### Environment

SM

### Version

- Component Version: Update from 8.7 to 8.8


### Rootcause

_No response_

### Solution ideas

_No response_

### Dev -> QA handover

- Resources:
- Versions to validate:
- Release version (in which version this feature will be released):


### Links

_No response_

### Automated Test Impact

_No response_

--- a/zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/serializer/ProtoBufSerializer.java
+++ b/zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/serializer/ProtoBufSerializer.java
@@ -537,7 +537,7 @@ public class ProtoBufSerializer
         pendingOperations);
   }
 
-  private Optional<RoutingState> decodeRoutingState(final Topology.RoutingState routingState) {
+  Optional<RoutingState> decodeRoutingState(final Topology.RoutingState routingState) {
     if (routingState.equals(Topology.RoutingState.getDefaultInstance())) {
       return Optional.empty();
     } else {
@@ -584,7 +584,7 @@ public class ProtoBufSerializer
     };
   }
 
-  private Topology.RoutingState encodeRoutingState(final RoutingState routingState) {
+  Topology.RoutingState encodeRoutingState(final RoutingState routingState) {
     return Topology.RoutingState.newBuilder()
         .setVersion(routingState.version())
         .addAllActivePartitions(routingState.requestHandling().activePartitions())

--- a/zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/serializer/ProtoBufSerializer.java
+++ b/zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/serializer/ProtoBufSerializer.java
@@ -66,6 +66,7 @@ import io.camunda.zeebe.dynamic.config.state.RoutingState.RequestHandling;
 import io.camunda.zeebe.util.Either;
 import java.nio.ByteBuffer;
 import java.time.Instant;
+import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Optional;
@@ -543,12 +544,15 @@ public class ProtoBufSerializer
       return Optional.of(
           new RoutingState(
               routingState.getVersion(),
-              decodeRequestHandling(routingState.getRequestHandling()),
+              decodeRequestHandling(
+                  routingState.getRequestHandling(), routingState.getActivePartitionsList()),
               decodeMessageCorrelation(routingState.getMessageCorrelation())));
     }
   }
 
-  private RequestHandling decodeRequestHandling(final Topology.RequestHandling requestHandling) {
+  private RequestHandling decodeRequestHandling(
+      final Topology.RequestHandling requestHandling,
+      @Deprecated(forRemoval = true) final List<Integer> activePartitionList) {
     return switch (requestHandling.getStrategyCase()) {
       case ALLPARTITIONS ->
           new RequestHandling.AllPartitions(requestHandling.getAllPartitions().getPartitionCount());
@@ -558,7 +562,15 @@ public class ProtoBufSerializer
               new TreeSet<>(
                   requestHandling.getActivePartitions().getAdditionalActivePartitionsList()),
               new TreeSet<>(requestHandling.getActivePartitions().getInactivePartitionsList()));
-      case STRATEGY_NOT_SET -> throw new IllegalArgumentException("Unknown request handling type");
+      case STRATEGY_NOT_SET -> {
+        // This fallback is used during the transition from 8.7 to 8.8, as RequestHandling was
+        // introduced in 8.8. It can be removed in 8.9
+        if (activePartitionList.isEmpty()) {
+          throw new IllegalArgumentException("Unknown request handling type");
+        } else {
+          yield new RequestHandling.AllPartitions(activePartitionList.size());
+        }
+      }
     };
   }
 
@@ -575,6 +587,7 @@ public class ProtoBufSerializer
   private Topology.RoutingState encodeRoutingState(final RoutingState routingState) {
     return Topology.RoutingState.newBuilder()
         .setVersion(routingState.version())
+        .addAllActivePartitions(routingState.requestHandling().activePartitions())
         .setRequestHandling(encodeRequestHandling(routingState.requestHandling()))
         .setMessageCorrelation(encodeMessageCorrelation(routingState.messageCorrelation()))
         .build();

--- a/zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/state/RoutingState.java
+++ b/zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/state/RoutingState.java
@@ -69,8 +69,7 @@ public record RoutingState(
 
   private static void validatePartitionCount(final int partitionCount) {
     if (partitionCount < 1) {
-      throw new IllegalArgumentException(
-          "Partition count must be greater than %d".formatted(Protocol.START_PARTITION_ID));
+      throw new IllegalArgumentException("Partition count must be greater than or equal to 1");
     }
   }
 

--- a/zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/state/RoutingState.java
+++ b/zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/state/RoutingState.java
@@ -12,6 +12,7 @@ import io.camunda.zeebe.protocol.Protocol;
 import java.util.HashSet;
 import java.util.Objects;
 import java.util.Set;
+import java.util.TreeSet;
 import java.util.function.UnaryOperator;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
@@ -64,6 +65,13 @@ public record RoutingState(
   public static RoutingState initializeWithPartitionCount(final int partitionCount) {
     return new RoutingState(
         1, new AllPartitions(partitionCount), new MessageCorrelation.HashMod(partitionCount));
+  }
+
+  private static void validatePartitionCount(final int partitionCount) {
+    if (partitionCount < 1) {
+      throw new IllegalArgumentException(
+          "Partition count must be greater than %d".formatted(Protocol.START_PARTITION_ID));
+    }
   }
 
   private static void validatePartitionId(final int partitionId) {
@@ -120,6 +128,8 @@ public record RoutingState(
       public ActivePartitions {
         Objects.requireNonNull(additionalActivePartitions);
         Objects.requireNonNull(inactivePartitions);
+        validatePartitionCount(basePartitionCount);
+
         for (final int activePartition : additionalActivePartitions) {
           validatePartitionId(activePartition);
           if (inactivePartitions.contains(activePartition)) {
@@ -136,8 +146,7 @@ public record RoutingState(
 
       @Override
       public Set<Integer> activePartitions() {
-        final var all =
-            new HashSet<Integer>(basePartitionCount + additionalActivePartitions.size());
+        final var all = new TreeSet<Integer>();
         all.addAll(allPartitions(basePartitionCount));
         all.addAll(additionalActivePartitions);
         all.removeAll(inactivePartitions);

--- a/zeebe/dynamic-config/src/main/resources/proto/topology.proto
+++ b/zeebe/dynamic-config/src/main/resources/proto/topology.proto
@@ -17,9 +17,10 @@ message ClusterTopology {
 }
 
 message RoutingState {
-  int64 version = 1;
-  RequestHandling requestHandling = 2;
+  int64  version = 1;
+  repeated sint32 activePartitions = 2 [deprecated=true];
   MessageCorrelation messageCorrelation = 3;
+  RequestHandling requestHandling = 4;
 }
 
 message RequestHandling {

--- a/zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/PersistedClusterConfigurationRandomizedPropertyTest.java
+++ b/zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/PersistedClusterConfigurationRandomizedPropertyTest.java
@@ -43,12 +43,6 @@ final class PersistedClusterConfigurationRandomizedPropertyTest {
     assertThat(updatedTopology).isEqualTo(persistedClusterTopology.getConfiguration());
     assertThat(PersistedClusterConfiguration.ofFile(topologyFile, serializer).getConfiguration())
         .usingRecursiveComparison()
-        // The type of generated `activePartitions` is `LinkedHashSet`, which AssertJ treats as an
-        // ordered collection. We are not interested in the order of the partitions, so we ignore
-        // it here.
-        .ignoringCollectionOrderInFields(
-            "routingState.value.requestHandling.additionalActivePartitions")
-        .ignoringCollectionOrderInFields("routingState.value.requestHandling.inactivePartitions")
         .isEqualTo(updatedTopology);
   }
 }

--- a/zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/serializer/ProtoBufSerializerPropertyTest.java
+++ b/zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/serializer/ProtoBufSerializerPropertyTest.java
@@ -34,12 +34,6 @@ final class ProtoBufSerializerPropertyTest {
     assertThat(decodedState.getClusterConfiguration())
         .describedAs("Decoded clusterTopology must be equal to initial one")
         .usingRecursiveComparison()
-        // The type of generated `activePartitions` is `LinkedHashSet`, which AssertJ treats as an
-        // ordered collection. We are not interested in the order of the partitions, so we ignore
-        // it here.
-        .ignoringCollectionOrderInFields(
-            "routingState.value.requestHandling.additionalActivePartitions")
-        .ignoringCollectionOrderInFields("routingState.value.requestHandling.inactivePartitions")
         .isEqualTo(clusterConfiguration);
   }
 }

--- a/zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/serializer/ProtoBufSerializerTest.java
+++ b/zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/serializer/ProtoBufSerializerTest.java
@@ -9,6 +9,7 @@ package io.camunda.zeebe.dynamic.config.serializer;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import com.google.protobuf.InvalidProtocolBufferException;
 import io.atomix.cluster.MemberId;
 import io.camunda.zeebe.dynamic.config.api.ClusterConfigurationChangeResponse;
 import io.camunda.zeebe.dynamic.config.api.ClusterConfigurationManagementRequest.AddMembersRequest;
@@ -23,9 +24,14 @@ import io.camunda.zeebe.dynamic.config.api.ClusterConfigurationManagementRequest
 import io.camunda.zeebe.dynamic.config.api.ClusterConfigurationManagementRequest.PurgeRequest;
 import io.camunda.zeebe.dynamic.config.api.ClusterConfigurationManagementRequest.ReassignPartitionsRequest;
 import io.camunda.zeebe.dynamic.config.api.ClusterConfigurationManagementRequest.RemoveMembersRequest;
+import io.camunda.zeebe.dynamic.config.protocol.Topology;
+import io.camunda.zeebe.dynamic.config.protocol.Topology.MessageCorrelation;
+import io.camunda.zeebe.dynamic.config.protocol.Topology.MessageCorrelation.HashMod;
+import io.camunda.zeebe.dynamic.config.protocol.Topology.RoutingState;
 import io.camunda.zeebe.dynamic.config.state.ClusterConfigurationChangeOperation.MemberLeaveOperation;
 import io.camunda.zeebe.dynamic.config.state.ClusterConfigurationChangeOperation.PartitionChangeOperation.PartitionJoinOperation;
 import io.camunda.zeebe.dynamic.config.state.MemberState;
+import io.camunda.zeebe.dynamic.config.state.RoutingState.RequestHandling;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -233,5 +239,26 @@ final class ProtoBufSerializerTest {
     final var decodedResponse =
         protoBufSerializer.decodeTopologyChangeResponse(encodedResponse).get();
     assertThat(decodedResponse).isEqualTo(topologyChangeResponse);
+  }
+
+  @Test
+  void shouldDecodeRoutingStateFrom87Correctly() throws InvalidProtocolBufferException {
+    final var routingState =
+        RoutingState.newBuilder()
+            .addAllActivePartitions(List.of(1, 2, 3))
+            .setMessageCorrelation(
+                MessageCorrelation.newBuilder()
+                    .setHashMod(HashMod.newBuilder().setPartitionCount(3))
+                    .build())
+            .build();
+    final var bytes = routingState.toByteArray();
+    final var decodedBytes =
+        protoBufSerializer.decodeRoutingState(Topology.RoutingState.parseFrom(bytes));
+    assertThat(decodedBytes)
+        .isPresent()
+        .hasValueSatisfying(
+            state -> {
+              assertThat(state.requestHandling()).isEqualTo(new RequestHandling.AllPartitions(3));
+            });
   }
 }

--- a/zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/util/ClusterTopologyDomain.java
+++ b/zeebe/dynamic-config/src/test/java/io/camunda/zeebe/dynamic/config/util/ClusterTopologyDomain.java
@@ -70,9 +70,9 @@ public final class ClusterTopologyDomain extends DomainContextBase {
 
   @Provide
   Arbitrary<RequestHandling> requestHandling() {
-    return Arbitraries.of(
-            ReflectUtil.implementationsOfSealedInterface(RequestHandling.class).toList())
-        .flatMap(Arbitraries::forType);
+    return Arbitraries.oneOf(
+        allPartitions().map(RequestHandling.class::cast),
+        activePartitions().map(RequestHandling.class::cast));
   }
 
   @Provide
@@ -83,8 +83,8 @@ public final class ClusterTopologyDomain extends DomainContextBase {
   @Provide
   Arbitrary<ActivePartitions> activePartitions() {
     final var basePartitionCount = Arbitraries.integers().between(1, 3);
-    final var activePartitions = Arbitraries.integers().between(4, 8).set();
-    final var inactivePartitions = Arbitraries.integers().between(9, 12).set();
+    final var activePartitions = Arbitraries.integers().between(4, 8).list().map(TreeSet::new);
+    final var inactivePartitions = Arbitraries.integers().between(9, 12).list().map(TreeSet::new);
 
     return Combinators.combine(basePartitionCount, activePartitions, inactivePartitions)
         .as(ActivePartitions::new);

